app/src/main/java/ch/epfl/sweng/zuluzulu/Structure/GPS.java

```diff
@@ -0,0 +1,173 @@
+package ch.epfl.sweng.zuluzulu.Structure;
+
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.location.Location;
+import android.location.LocationListener;
+import android.location.LocationManager;
+import android.os.Bundle;
+import android.support.v4.content.ContextCompat;
+import android.util.Log;
+import android.widget.Toast;
+
+import static android.content.Context.LOCATION_SERVICE;
+
+public final class GPS implements LocationListener {
+
+    private static volatile GPS instance = null;
+
+    public static final int MY_PERMISSIONS_REQUEST_LOCATION = 1;
+    private static final long MIN_DISTANCE_TO_REQUEST_LOCATION = 1; // In meters
+    private static final long MIN_TIME_FOR_UPDATES = 1000; // 1 sec
+    private static final int TWO_MINUTES = 1000 * 60 * 2; // 2 min
+
+    private Context context;
+    private Location location;
+    private LocationManager locationManager;
+
+    private boolean isWorking = false;
+
+    private GPS(Context context) {
+        super();
+        this.context = context.getApplicationContext();
+    }
+
+    public static GPS getInstance(Context context) {
+        if (instance == null) {
+            synchronized (GPS.class) {
+                if (instance == null) {
+                    instance = new GPS(context);
+                }
+            }
+        }
+        instance.setContext(context);
+        return instance;
+    }
+
+    /**
+     * Start asking for location updates and return true if succeed
+     */
+    public boolean start() {
+        if (ContextCompat.checkSelfPermission(context, android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
```

Comment:
You can still refactor it a little, for instance 
```
 if (ContextCompat.checkSelfPermission(context, android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
 Toast.makeText(context, "Permission to GPS not granted", Toast.LENGTH_SHORT).show();
            return false;
        }
```


----> if(this.checkPermission())

Reply:
Android Studio will complain if I wrap the checkSelfPermission. It seems to be a special case.
But yeah I am still working on refactoring what I can.

Reply:
Oh ok, I didn't knew about this requirement.  Good job :+1: